apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -21,28 +32,108 @@ def get_file_names_as_array(raw_file_names: str) -> list[str]:
-# file_names attribute
-def get_answer_name(file_names: str) -> str:
-    return "_file_upload_{}".format(
-        hashlib.sha1(file_names.encode("utf-8")).hexdigest()
+def match_regex_with_files(
+    regex_patterns: list[str], files_names: set[str], limit_1: bool
+) -> tuple[list[str], list[str]]:
+    """
+    Take a list of regexes and a list of file names and matches them 1:n or 1:1, depending on limit_1 parameter
+    Returns a tuple of matched file names and unmatched patterns
+    """
+    unmatched_patterns = regex_patterns.copy()
+    remaining_files = files_names.copy()
+    for pattern in regex_patterns:
+        regex = re.compile(pattern, re.IGNORECASE)
+        for file in remaining_files.copy():
+            if regex.match(file) and (not limit_1 or pattern in unmatched_patterns):
+                if pattern in unmatched_patterns:
+                    unmatched_patterns.remove(pattern)
+                remaining_files.remove(file)
+
+    return [f for f in files_names if f not in remaining_files], unmatched_patterns
+
+
+def glob_to_regex(glob_pattern: str) -> str:
+    """Translate a glob pattern into a regex that can be handled consistently across Python and JS.
+
+    Returns:
+        The regex as a string
+
+    Raises:
+        ValueError: If the glob pattern is invalid
+    """
+    if "/" in glob_pattern:
+        raise ValueError(
+            f"The file name pattern {glob_pattern} contains a '/' character, which is not allowed in file names."
+        )
+    return fnmatch.translate(glob_pattern).replace("(?s:", "^").replace(r")\Z", "$")
+
+
+def get_answer_name(
+    file_names: str | None,
+    optional_file_names: str | None = None,
+    file_patterns: str | None = None,
+    optional_file_patterns: str | None = None,
+) -> str:
+    """
+    Compute the unique identifer of a pl-file-upload element, which is the SHA1 hash of all its
+    file name attributes
+    """
+    # Using / as separator as the only character guaranteed not to appear in file names
+    combined_name = (
+        (file_names or "")
+        + ("/" + optional_file_names if optional_file_names else "")
+        + ("//" + file_patterns if file_patterns else "")
+        + ("////" + optional_file_patterns if optional_file_patterns else "")
     )
 
+    return "_file_upload_" + hashlib.sha1(combined_name.encode("utf-8")).hexdigest()
 
-def add_format_error(data: pl.QuestionData, error_string: str) -> None:
+
+def add_format_error(
+    answer_name: str, data: pl.QuestionData, error_string: str
+) -> None:
+    # Adding format errors to both answer_name and "_files" keys for display next to this
+    # element and in submissions
     pl.add_files_format_error(data, error_string)
 
+    if answer_name not in data["format_errors"]:
+        data["format_errors"][answer_name] = []
+    data["format_errors"][answer_name].append(error_string)
+
 
 def prepare(element_html: str, data: pl.QuestionData) -> None:
```

Comment:
Nit: let's consider validating the regular expressions here in `prepare()`, specifically to check that there aren't any slashes.